### PR TITLE
CMake: Add code for Xcode generation

### LINKED ISSUE #219: Can SVT-AV1 generate xcode project ?
Hi SVT experts, 

I want to debug the SVT-AV1 code by using the xcode in MacOS, so could you please tell me how to generate the xcode project for SVT-AV1 ?

### PR DESCRIPTION
**Pull request by 1480c1**
Based off code from x265's CMakeLists

Required some "Hackish" code due to the unique structure.

Closes #219

### CODE REVIEW

**Comment by 1480c1** [comment]
Not complete, ment to open as draft

**Comment by 1480c1** [comment]
Probably not WIP anymore. Works with Xcode, MSVC, and Unix Makefiles

**Comment by joelsole** [comment]
This is useful!
Could you add the corresponding instructions to the README?

**Comment by 1480c1** [comment]
> This is useful!
> Could you add the corresponding instructions to the README?

Sure!

**Comment by hassount** [comment]
@1480c1 
Thank you for the contribution and stating the source, could you please share with us the license header as there might be a license conflict here.

**Comment by 1480c1** [comment]
I think based off of http://x265.org/, it's GPLv2

**Comment by 1480c1** [comment]
And although I say it's code from x265, I'm only loosely basing the logic off of what they did in their CMakeLists

**Comment by 1480c1** [comment]
I think this pr is ready for review,

**Comment by joelsole** [comment]
Not working for me: 

....
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't open file: /Users/jsole/Documents/code/SVTAV1/ovc_SVT-AV1/Build/Source/Lib/Common/ASM_AVX2/svt-av1.build/Debug/COMMON_ASM_AVX2.build/Objects-normal/undefined_arch/wiener_convolve_avx2.o (No such file or directory)
Command Libtool failed with a nonzero exit code

Maybe I missed something. Could you add info at README?

**Comment by 1480c1** [comment]
Theoretically, all you would need to run is
``` bash
cd Build
cmake .. -G"Xcode" -DCMAKE_BUILD_TYPE=Release
cmake --build .
```

Currently, the build.sh does not support additional generators, but either way, the error seems weird to me since I have not touched anything in the AVX2 folder. Will try to see if I can figure out what's wrong

**Comment by joelsole** [comment]
I was doing something different. With your commands, I get the error below. Note the highlighted path.


** BUILD FAILED **


The following build commands failed:
        Libtool /Users/jsole/Documents/code/SVTAV1/ovc_SVT-AV1/Bin**/Release/Debug**/libSvtAv1Enc.a normal x86_64

**Comment by 1480c1** [comment]
That's one problem with IDE generators (Xcode and Visual Studio), their CMAKE_BUILD_TYPE is "Release;Debug" and their are multi-config generators. As such, they do not mesh well when using CMAKE_BUILD_TYPE to output something in the CMakeLists

**Comment by 1480c1** [comment]
I probably will remove the xcode travis build job once I confirm it all works

**Comment by joelsole** [comment]
> I was doing something different. With your commands, I get the error below. Note the path ../Bin/Release/Debug/...
> 
> ** BUILD FAILED **
> 
> The following build commands failed:
> Libtool /Users/jsole/Documents/code/SVTAV1/ovc_SVT-AV1/Bin/Release/Debug/libSvtAv1Enc.a normal x86_64

Still not building for me. @andrey-norkin , could you check it?

**Comment by 1480c1** [comment]
There doesn't seem to be a function for `eb_aom_subtract_block_sse2`

https://travis-ci.com/1480c1/SVT-AV1/jobs/236647122#L2374

**Comment by EwoutH** [comment]
@1480c1 Could you resolve conflicts and rebase?

**Comment by 1480c1** [comment]
done

**Comment by EwoutH** [comment]
The new Xcode job currently fails

**Comment by 1480c1** [comment]
I was missing an asm file in the cmakelists

**Comment by EwoutH** [comment]
Nice job!

**Comment by 1480c1** [comment]
Should the travis job be debug or release?

**Comment by EwoutH** [comment]
My preference would be Release (to catch optimization errors, and that's what most users will be using), but both is ok. There are Debug builds already elsewhere in Travis.

**Comment by JackZhouVCD** [comment]
Thanks for adding this code! I applied this PR and it worked on 10.14.6 (Mojave). However, I have a question. The list of projects I'm seeing on Xcode is not the same as the list of projects I see if I'm using VS2017:

|VS2017|Xcode|
|-|-|
|vs2017projects|Screen Shot 2019-10-09 at 12 11 46 PM|

Is there a reason for these extra projects to be there on Xcode?

**Comment by 1480c1** [comment]
It has to create a separate "project" for each asm file, even though they're actually just a one line script to run nasm/yasm

**Comment by 1480c1** [comment]
Each of the additional project should actually just be an asm file. In the normal branch (Everything excluding xcode), all of the asm files are processed fine, but it seems cmake lacks proper support for generating xcode project files that automatically process nasm files, leading to where they show up in the project files, but are labeled as generic "Source Code" and not even processing correctly even with "nasm Source Code" applied

**Comment by 1480c1** [comment]
On hold until I can decide how I want to add asm libraries for xcode builds

**Comment by hassount** [comment]
can we rebase this ? let's try to get this working by the code freeze otherwise we can close it

**Comment by 1480c1** [comment]
Sure, I can try to work on it a bit tomorrow

**Comment by 1480c1** [comment]
Do not merge for now since I still haven't actually made sure it works

**Comment by hassount** [comment]
> Do not merge for now since I still haven't actually made sure it works

Sure, no rush, we can do this during the cleanup phase

**Comment by 1480c1** [comment]
Will comment out the xcode job after it's done since that job takes way too long to keep

**Comment by 1480c1** [comment]
@JackZhouVCD and @joelsole, can you check to make sure that the xcode project looks fine?

I've set it up so that for Xcode generation, cmake will generate the object files during the configure stage and add the objects as sources to libraries linking the interface library,  `common_lib`. So you should see the object files inside the Xcode viewer instead of the asm files since I can't get it to work with the asm files properly without needing to add `set_source_files_properties(${SIMD_OBJS} PROPERTIES GENERATED TRUE)` everytime we link them.

**Comment by 1480c1** [comment]
@hassount, is the amount of commits fine, or do you want me to squash them?

**Comment by hassount** [comment]
@1480c1 
yeah let's squash them to 1 or 2 at most, thank you

**Comment by hassount** [comment]
@joelsole can you check if this is working now on your end ?

**Comment by 1480c1** [comment]
> @1480c1
> yeah let's squash them to 1 or 2 at most, thank you

Done

**Comment by JackZhouVCD** [comment]
> @JackZhouVCD and @joelsole, can you check to make sure that the xcode project looks fine?

Yup! Looks good on my end

> I've set it up so that for Xcode generation, cmake will generate the object files during the configure stage and add the objects as sources to libraries linking the interface library, `common_lib`. So you should see the object files inside the Xcode viewer instead of the asm files since I can't get it to work with the asm files properly without needing to add `set_source_files_properties(${SIMD_OBJS} PROPERTIES GENERATED TRUE)` everytime we link them.

Nice, the project list looks a lot more clean now

**Comment by hassount** [comment]
@JackZhouVCD are you able to build on xcode ?

**Comment by JackZhouVCD** [comment]
> @JackZhouVCD are you able to build on xcode ?

yup, building the encoder works

**Comment by joelsole** [comment]
That's great. Is the README reflecting all the steps you took to make it work?

**Comment by 1480c1** [comment]
you should be able to use the cmake commands directly (`cmake .. -GXcode [...]`) or use the build script (`./Build/linux/build.sh -g Xcode`), although the build script doesn't have support for only generating yet and will autobuild. I might want to add a `no-build` option to it similar to `build.bat` in the future

File: README.md [hunk 1]
```diff
@@ -86,8 +86,18 @@ The SVT-AV1 Encoder library supports the x86 architecture
...
+    cd Build/linux
+    ./build.sh <release | debug>
+    ```
+
+    To build using Xcode:
+
+    ``` bash
+    cd Build/linux
+    ./build.sh <release | debug> -g Xcode
+    ```
```
Comment (1480c1) [inline README.md]: The commands to build a xcode project and build it is here

**Comment by andrey-norkin** [comment]
It seems to be working for me, able to compile according to the instructions. 

I also tried generating, opening, and compiling the project in the XCode IDE and it worked by doing the following. 
'cmake .. -G Xcode'
However, when compiling this way, a number of warnings are generated, on the following type:
"ld: warning: directory not found for option '-L/Users/.../SVT-AV1/Source/Lib/Common/C_DEFAULT/Release'"

We should try to get rid of them too.

**Comment by 1480c1** [comment]
> However, when compiling this way, a number of warnings are generated, on the following type:
> "ld: warning: directory not found for option '-L/Users/.../SVT-AV1/Source/Lib/Common/C_DEFAULT/Release'"

I'll try to look for what's causing that, but as far as I can tell, it does not seem to be coming from our CMakeLists.

Also, you can use backticks `` ` `` to separate out `certain stuff` instead of using single quotes

**Comment by hassount** [comment]
@1480c1 can you please rebase this so that we can merge it ?

**Comment by 1480c1** [comment]
Done

**Comment by hassount** [comment]
thnx Chris, let's give it a chance to pass the tests and merge it

**Comment by hassount** [comment]
@1480c1 
sorry, would you please rebase again ?

**Comment by 1480c1** [comment]
Done

**Comment by 1480c1** [comment]
Going to wait until I can figure out what happened in https://github.com/OpenVisualCloud/SVT-AV1/pull/917 before rebasing

**Comment by 1480c1** [comment]
@andrey-norkin, will there be a `Enc-common separation part 4`?

**Comment by andrey-norkin** [comment]
Yes, there will be Part 4. The SIMD folders in the common library would be back and dependencies of the decoder library on encoder removed. Part 3 was an intermediate step to facilitate further work on separation of the libraries.

**Comment by 1480c1** [comment]
Okay, I will wait until part 4 before continuing since it's going to be hard to maintain this with constant cmake changes

**Comment by andrey-norkin** [comment]
> Okay, I will wait until part 4 before continuing since it's going to be hard to maintain this with constant cmake changes

Feel free to proceed with this PR. The separation has been finished, and the dependencies between the encoder, decoder, and common libraries are back to what they used to be.

**Comment by 1480c1** [comment]
> > Okay, I will wait until part 4 before continuing since it's going to be hard to maintain this with constant cmake changes
> 
> Feel free to proceed with this PR. The separation has been finished, and the dependencies between the encoder, decoder, and common libraries are back to what they used to be.

Will do once I have some free time

**Comment by ePirat** [comment]
I had a quick look and the commit seems to have a lot of unrelated changes that should either be different PRs altogether or at least split into multiple commits.

**Comment by 1480c1** [comment]
> I had a quick look and the commit seems to have a lot of unrelated changes that should either be different PRs altogether or at least split into multiple commits.

Yeah, in a different branch that I was working on rebasing this, I did have them split with actual proper commit messages. I can go ahead and cherry pick them into an actual PR

**Comment by 1480c1** [comment]
@ePirat can you test this pr. I do not have a working mac with Xcode to check if it works.

**Comment by ePirat** [comment]
Yep it works 👍

**Comment by ePirat** [comment]
Please create the dummy.c in the build dir or add it to the `.gitignore`

**Comment by hassount** [comment]
@1480c1 thank you very much for your help with this.
@joelsole @andrey-norkin is this good on your end ?

**Comment by 1480c1** [comment]
> Please create the dummy.c in the build dir or add it to the `.gitignore`

Moved the generated dummy.c file into the project_binary_dir

**Comment by 1480c1** [comment]
> @1480c1 thank you very much for your help with this.
> @joelsole @andrey-norkin is this good on your end ?

Ping?

**Comment by ePirat** [comment]
@hassount Would be good if this can be reviewed/merged soon as I believe it was relatively important and I don't want to make PRs that conflict with this. (Which would happen when I continue work on #1200)

**Comment by andrey-norkin** [comment]
Xcode builds but gives a number of warnings of the following type:
ld: warning: directory not found for option '-L/.../SVT-AV1/Source/Lib/Common/ASM_SSE2/Debug'

I suggest to merge though anyway so that the project is at least building

**Comment by hassount** [comment]
Is there anyway we could add a ci job to build with xcode to make sure we don’t break this ?

**Pull request merged**